README
xvnkb  - Vietnamese keyboard input for X-Window

1. Description
    xvnkb is a Vietnamese keyboard input for X-Window. It provides a
    useful way for editing Vietnamese on X-Window environment with popular
    input methods and charsets. The software is released under the GNU 
    General Public License.

2. Features
    - Runs as a small X application in the background without any interfere
      with others.
    - Can work well with most X applications included: editors, browsers,
      office suites, ...
    - Supported Vietnamese input methods: VNI, TELEX, VIQR.
    - Supported Vietnamese charsets: TCVN3, VISCII, VPS, VNI, VIQR, Unicode
      (UTF-8).
    - Switch on/off input mode with shortcut key or with left mouse click on 
      the running icon.
    - Has been successfully tested with XFree86 on Linux, FreeBSD, OpenBSD, 
      and hope to work on other UNICES (anyone can test please contact me).

3. User Guide
3.1. Install
    The install script will help you to do all things:

    [xvnkb2-source]$ ./configure
    [xvnkb2-source]$ make install

    or

    [xvnkb2-source]$ ./configure
    [xvnkb2-source]$ make
    [xvnkb2-source]$ ./install

    and then follow the instructions of the install script.

    See the INSTALL file for more information.

3.2. Launch
    To control the status of xvnkb core (xvnkb.so), launch xvnkb gui

    $ xvnkb

    A small icon of xvnkb will appear on the screen and display the status of
    xvnkb core (input mode), which can be changed by using mouse:
    - Left button: Switch off the Vietnamese typing mode or switch on the previous
      input method,
    - Right button: Activate menu for configuring xvnkb,
    - Middle button: Exit.

    Pesonal configurations (status of current input method, charset,...) are 
    stored in file $HOME/.xvnkbrc and will be restored in the next session of xvnkb.

3.3. Manually install & launch
    In the case you want to do it manually:

    [xvnkb2-source]$ ./configure
    [xvnkb2-source]$ make
    [xvnkb2-source]$ cp xvnkb.so.<version> /path/to/your/lib/folder
    [xvnkb2-source]$ cp xvnkb /path/to/your/bin/folder

    $ export LD_PRELOAD=/path/to/your/lib/folder/xvnkb.so.<version>
    $ <launch xvnkb to control>
    $ <launch your apps here>

3.4. Function
    With your default LANG, xvnkb can help you to input following charsets:
    tcvn, vni, viscii, vps,... but maybe not unicode. So, when you want to input
    unicode, you have to set LANG=UTF-8.

    $ export LANG=UTF-8       # or LANG=en_US.UTF-8 or {something}.UTF-8
    $ <launch your apps here>

    * NOTICE:
    xvnkb works fine with all other charsets when LANG=*UTF-8 is set.

3.5. Hints for advanced users (who don't want to use GUI :)):
    Make core only with default input method = viqr

    $ make core

    Make core with debug (viqr)

    $ make core_debug

    Make core with user defined input method

    $ make core CFLAGS="-DVK_CORE_METHOD=<N>"

    where N: 1=vni, 2=telex, 3=viqr

    Also, you can change input method by setting environment variable
    VK_CORE_METHOD

    $ export VK_CORE_METHOD=1 # vni input method
    $ <launch apps>

    In this case GUI has no effect on the core. Use Alt-Shift to switch
    on/off Vietnamese input method (default is off).

    See also 3.3

4. Limits
    - Installation script works only if bash exists.


Thank you for using xvnkb!!! Have fun!!!

Dao Hai Lam <daohailam@yahoo.com>


ChangeLog
0.2.11:
	* 2012/05/25:
	- Fixed backspace bug in Firefox
	- Using static to avoid conflict variables with other apps
0.2.10:
	* 2009/06/25:
	- Replace "usleep" with "select" to prevent laptop battery consumption
	- Use constants for docking status
	- Set default input method, charset for the first time
	- Fixed double free in resource file
0.2.9:
	* 2006/12/07:
	- Use Unicode (UTF-8) as default charset
	- Set some default info (docking, using method) & save for the first time
	- Allow input Viet Nam Dong sign (telex: dd_, vni: d9_, viqr: dd_)
	* 2006/02/26:
	- Use smaller size for default font
	- Implement systray size calculation (work around for some systray
	  non-compliant with freedesktop specs)
	- Use type-casting to avoid some warning messages when compile
	- Cleanup & fix bugs
	* 2005/07/18:
	- No need to check LOCALE anymore, almost distro are ready for UTF-8 locale
	* 2005/04/01:
	- Init default resource
	* 2004/01/24:
	- Fix telnet segmentation fault
	- Remove fast VNI
0.2.8:
	* 2004/01/17:
	- Release
	- Simplify & fix some bugs in new_install.sh/new_uninstall.sh
	- Use size hint in docking mode only (fix bug when drawing floating icon)
	* 2004/01/16:
	- Add new_install.sh/new_uninstall.sh scripts, new xvnkb.spec
	- Add old_install & old_uninstall entries: old install/uninstall style
	* 2004/01/13:
	- Fix thinko in size calcation of Message box
	- Display message box instead of exit if error occurs on setting locale
	* 2004/01/12:
	- Checking LD_PRELOAD, LANG & locale information when xvnkb (GUI) start
	- Add message box
	* 2004/01/11:
	- Undocking when systray has been removed & redocking when systray reappears
	* 2004/01/07:
	- Add dock/undock capabilities
	- Save docking information to resource file
	* 2004/01/06:
	- GUI - dockable to systray
	- CORE - force setting GTK_IM_MODILE=xim (work around for gtk2/gnome2)
	* 2003/11/20:
	- Name correction for Telex method, ABC instead of Vietkey (much better :))
	  -- Rename telex-vkstyle.h => telex-abcstyle.h
	  -- Usage: --use-abcstroke as option of configure script to enable
	- Changed configure script's option names: --use-{ext,pro,abc}stroke
	- Changed some marcro names: VK_USE_{EXT,PRO,ABC}STROKE
	- Cleanup some stupid codes
	- Fixed more bugs/thinko in extended Telex mode
	  -- Usage: --use-extstroke as option of configure script to enable
	  -- Rule: ] => u+, [ => o+, } => U+, { => O+
	           w => u+, W => U+
	- Add ProTelex mode (Programmer Telex mode - use by myself :))
	  -- Usage: --use-prostroke as option of configure script to enable
	  -- Rule: [ => u+, ] => o+, { => U+, } => O+
	           w => u+, W => U+
	* 2003/11/18:
	- Fixed thinko in Shift/Capslock case => Fixed bug in VNI input method
	- Add more extra keystroke for all input methods
	* 2003/11/15:
	- Use different word separators for each input method
	- Add Makefile.sunos => Makefile.linux (use GNU make syntax temporarily)
	- Add flags supporting for linking shared library on SunSolaris
	- Support some extra Telex keystrokes
	  -- Use --use-xstroke as option of configure script to enable
	  -- Rule: [ => u+, ] => o+, { => U+, } => O+
	           w => u+, W => U+ (where 'w' or 'W' is start character of a word)
	- Use Telex standard key stroke as default - Vietkey style is option
	  -- Use --use-vkstroke as option of configure script to replace standard
	* 2003/11/14:
	- Remove Bash syntax in configure
	- Add checking for uchar, ushort, uint, ulong
	* 2003/10/02:
	- Create softlink to particular Makefile based on Platform info (configure)
	- Add Makefile for *BSD, separate common code & dependent code
	- Allow using "Windows" key as hotkey prefix
	- Cleanup some code in xresource.c
	* 2003/XX/XX:
	- Add menu option to disable/enable spellcheck
	- Fixed compilation on FreeBSD
	- Fixed configure/Makefile (checking dl, xft libs)
	- Fixed crashing on KDE 3.x (cvs version)
	- Fixed error with determination of focus window
	  -- Very rarely, found in Links2.1pre9 graphics only (at the moment)
	- Fixed spell checking with "gio<...>"
	- Fixed compilation bug with Debian gcc 2.95.4
	- Updated documents

0.2.7:
	- Fixed Hotkey recording
	- Fixed configure script for XFT detecting
	  -- Looking for xft-config script
	- Enable spell checking as default in configure script
	  -- Disable with --no-spellcheck option
	- Display Vietnamese menu w/o using XFT
	  -- use XmbDrawString

0.2.6:
	- Use XFT (support Vietnamese interface & antilias)
	  -- Auto detect, disable with option --no-xft
	- Remove gtk2 module
	  -- Move xvnkb_gtk.c to "obsolete" folder.
	  -- Remove option --with-gtk2 in configure script
	- Add Event handler for all windows (event.c)
	- Add "Select hotkey" dialog/add resource entry for hotkey
	- Change popup menu style (cascade)
	- Fixed more spell check
	- Fixed tools/xvnkb_ctrl

	- Update documents for Gtk/Gnome 2.x users.
	  -- Gtk/Gnome 2.x apps can use X-Input-Method with xvnkb (see README)

0.2.5:
	- Fix & update documents (locale)
	- Add more options for shared libs compilation
	- Add support for gtk2 (xvnkb_gtk.c)
	  -- Add option --with-gtk2 for configure script
	  -- Usage: $ ./configure --with-gtk2
	- Add more functions for installation script
	  -- Detect and help user create UTF-8 locale
	- Add more scripts for installation (localesetup.sh, xvnkb.sh)
	  -- localesetup.sh: help user creating locale
	  -- xvnkb.sh: a shell script which help user set LD_PRELOAD, LANG automatically
	- Code cleanup

0.2.4:
	- Smart Vietnamese input (auto regconize Foreign languages word): spell checking
	  -- Add option --spell-check for configure script
	  -- Usage: $ ./configure --spell-check
	- Rewrite xvnkb_ctrl
	  -- use xconfig source for load/save configuration
	  -- add option -t for toggle status
	- Optimize and cleanup code

	- Add version info used for all scripts (file VERSION)
	- Add version number to xvnkb.so. In this release it will be xvnkb.so.0.2.4
	- Safe install/uninstall scripts, add version info to all xvnkb scripts

menu.c
/*----------------------------------------------------------------------------*/
/*  menu.c                                                                    */
/*----------------------------------------------------------------------------*/
/*  copyright         : (C) 2002 by Dao Hai Lam                               */
/*                      VISC Software & Security Consultant Company           */
/*                      Hall 3, Quang Trung Software City                     */
/*                      Tan Chanh Hiep Ward, District 12,                     */
/*                      Ho Chi Minh city, VIETNAM                             */
/*  website           : http://www.visc-network.com                           */
/*  email             : lam@visc-network.com                                  */
/*  last modify       : Tue, 13 Aug 2002 03:14:07 +0700                       */
/*----------------------------------------------------------------------------*/
/*                                                                            */
/*   This program is free software; you can redistribute it and/or modify     */
/*   it under the terms of the GNU General Public License as published by     */
/*   the Free Software Foundation; either version 2 of the License, or        */
/*   (at your option) any later version.                                      */
/*                                                                            */
/*----------------------------------------------------------------------------*/
#include "data.h"
#include "flash.h"
#include "event.h"
#include "mainwin.h"
#include "systray.h"
#include "menu.h"
/*----------------------------------------------------------------------------*/
#ifdef VK_CHECK_SPELLING
#define DOCK_MSG_IDX	3
#else
#define DOCK_MSG_IDX	2
#endif
/*----------------------------------------------------------------------------*/
typedef struct vk_menu_item {
	int x, y;
	char *name;
	int type;
	int *vptr;
	int value;
	int length;
	int width;
	struct vk_menu_item *next;
	struct vk_menu *menu;
} vk_menu_item;
/*----------------------------------------------------------------------------*/
typedef struct vk_menu {
	int x, y;
	int w, h;
	int tx, ty;
	char *name;
	int sidex;
	vk_menu_item *sitem;
	vk_menu_item *items;

	Window win;
	GC gc;
} vk_menu;
/*----------------------------------------------------------------------------*/
#define MENU_PADDING	4
/*----------------------------------------------------------------------------*/
static char *vk_dock_message[][2] = {
	{"Docking to systray", "Undocking"},
	{"Đặt vào khay", "Gỡ khỏi khay"}
};
/*----------------------------------------------------------------------------*/
vk_menu_item vk_menu_data[][25] =
{
	{
	{ 0, 0, "About",				-2 },
	{ 0, 0, "Miscellaneous",		 0 },
#ifdef VK_CHECK_SPELLING
	{ 0, 0, "Check spelling",		 7, &vk_spelling, 1 },
#endif
	{ 0, 0, NULL,					 8, NULL },
	{ 0, 0, "Set hotkey",			 5, NULL },
	{ 0, 0, "Set interface font",	 6, NULL },
	{ 0, 0, "Input method",			 0 },
	{ 0, 0, "Off",					 1, &vk_method, VKM_OFF },
	{ 0, 0, "Vni",					 1, &vk_method, VKM_VNI },
	{ 0, 0, "Telex",				 1, &vk_method, VKM_TELEX },
	{ 0, 0, "Viqr",					 1, &vk_method, VKM_VIQR },
	{ 0, 0, "Charset",				 0 },
	{ 0, 0, "Unicode",				 3, &vk_charset, VKC_UTF8 },
	{ 0, 0, "Tcvn3",				 3, &vk_charset, VKC_TCVN },
	{ 0, 0, "Vni",					 3, &vk_charset, VKC_VNI },
	{ 0, 0, "Viqr",					 3, &vk_charset, VKC_VIQR },
	{ 0, 0, "Viscii",				 3, &vk_charset, VKC_VISCII },
	{ 0, 0, "Vps",					 3, &vk_charset, VKC_VPS },
	{ 0, 0, "Interface",			 0 },
	{ 0, 0, "English",				 4, &vk_interface, VK_ENGLISH },
	{ 0, 0, "Vietnamese",			 4, &vk_interface, VK_VIETNAMESE },
	{ 0, 0, "Exit",					-1 },
	{ 0, 0, NULL,					 0 }
	},
	{
	{ 0, 0, "Thông tin",			-2 },
	{ 0, 0, "Linh tinh",			 0 },
#ifdef VK_CHECK_SPELLING
	{ 0, 0, "Kiểm tra từ",			 7, &vk_spelling, 1 },
#endif
	{ 0, 0, NULL,					 8, NULL },
	{ 0, 0, "Chọn phím bật/tắt",	 5, NULL },
	{ 0, 0, "Chọn font giao diện",	 6, NULL },
	{ 0, 0, "Kiểu gõ",				 0 },
	{ 0, 0, "Tắt",					 1, &vk_method, VKM_OFF },
	{ 0, 0, "Vni",					 1, &vk_method, VKM_VNI },
	{ 0, 0, "Telex",				 1, &vk_method, VKM_TELEX },
	{ 0, 0, "Viqr",					 1, &vk_method, VKM_VIQR },
	{ 0, 0, "Bảng mã",				 0 },
	{ 0, 0, "Unicode",				 3, &vk_charset, VKC_UTF8 },
	{ 0, 0, "Tcvn3",				 3, &vk_charset, VKC_TCVN },
	{ 0, 0, "Vni",					 3, &vk_charset, VKC_VNI },
	{ 0, 0, "Viqr",					 3, &vk_charset, VKC_VIQR },
	{ 0, 0, "Viscii",				 3, &vk_charset, VKC_VISCII },
	{ 0, 0, "Vps",					 3, &vk_charset, VKC_VPS },
	{ 0, 0, "Giao diện",			 0 },
	{ 0, 0, "Tiếng Anh",			 4, &vk_interface, VK_ENGLISH },
	{ 0, 0, "Tiếng Việt",			 4, &vk_interface, VK_VIETNAMESE },
	{ 0, 0, "Thoát",				-1 },
	{ 0, 0, NULL,					 0 }
	}
};
/*----------------------------------------------------------------------------*/
vk_menu menu = {
	0, 0, 0, 0, 0, 0,
	"xvnkb  "VERSION,
	0, NULL, NULL,
	0, 0
};
/*----------------------------------------------------------------------------*/
extern void VKDrawIcon();
extern void VKIconLoadInterface();
extern void VKCreateFlash();
extern void VKDestroyFlash();
extern void VKShowHotkeyWindow();
extern void VKDrawHotkeyWindow();
/*----------------------------------------------------------------------------*/
void MenuProcess(XEvent *event, void *data);
/*----------------------------------------------------------------------------*/
void CreateMenuWindow(vk_menu *m, int w, int h)
{
	m->w = w + 36;
	m->h = h * (vk_text_height + 4) + MENU_PADDING;
	m->tx = 12;
	m->ty = vk_text_ascent + 4;
	m->win = XCreateWindow(display, root,
					m->x, m->y, m->w, m->h, 0,
					DefaultDepth(display, screen), InputOutput, visual,
					vk_attribmask, &vk_attrib);
	XSelectInput(display, m->win, MAIN_EVENT_MASKS);

	m->gc = XCreateGC(display, m->win, 0, 0);
	XDefineCursor(display, m->win, vk_hand);
#ifndef USE_XFT
	XSetFont(display, m->gc, vk_font->fid);
#endif
	VKRegisterEvent(m->win, MenuProcess, m);
}
/*----------------------------------------------------------------------------*/
void CalculateItemWidth(vk_menu_item *current, vk_menu_item *item)
{
	char *s = item->name;
#ifdef USE_XFT
	XGlyphInfo fi;
	XftTextExtentsUtf8(display, vk_font, (uchar *)s, item->length = strlen(s), &fi);
#else
	XRectangle fi;
	XmbTextExtents(vk_fontset, s, item->length = strlen(s), 0, &fi);
#endif
	item->width = fi.width;
	if( current ) current->next = (struct vk_menu_item *)item;
}
/*----------------------------------------------------------------------------*/
void CalculateItemPosition(vk_menu *m)
{
	int h = m->ty + vk_text_height + 4;
	vk_menu_item *item = m->items;
	while( item && item->name ) {
		item->x = m->tx;
		item->y = h;
		item = (vk_menu_item *)item->next;
		h += (vk_text_height + 4);
	}
}
/*----------------------------------------------------------------------------*/
vk_menu_item *CreateSubMenu(vk_menu_item *item, vk_menu_item *child)
{
	int w, h;
	vk_menu_item *citem = NULL;
	vk_menu *m = item->menu = (vk_menu *)calloc(1, sizeof(vk_menu));

	CalculateItemWidth(citem, item);
	h = 1; w = item->width;
	m->name = item->name;
	m->items = child;
	while( child && child->type>0 ) {
		CalculateItemWidth(citem, child);
		if( w<child->width )
			w = child->width;
		citem = child;
		child++;
		h++;
	}
	citem->next = NULL;
	CreateMenuWindow(m, w, h);
	CalculateItemPosition(m);

	return child;
}
/*----------------------------------------------------------------------------*/
void VKCreateMenuWindow()
{
	int h = 1, w;
	char *s = menu.name;
	vk_menu_item *citem = NULL;
	vk_menu_item *item = vk_menu_data[vk_interface];
#ifdef USE_XFT
	XGlyphInfo fi;
	XftTextExtentsUtf8(display, vk_font, (uchar *)s, strlen(s), &fi);
#else
	XRectangle fi;
	XmbTextExtents(vk_fontset, s, strlen(s), 0, &fi);
#endif
	w = fi.width;

	VKUpdateDockingMessage();
	while( item && item->name ) {
		switch( item->type ) {
			case 0:
				if( !citem )
					menu.items = citem = item;
				else {
					citem->next = (struct vk_menu_item *)item;
					citem = item;
				}

				item = CreateSubMenu(item, &item[1]);
				if( w<citem->width ) w = citem->width;
				break;
			default:
				CalculateItemWidth(citem, item);
				if( w<item->width )	w = item->width;
				if( !citem ) menu.items = item;
				citem = item;
				item++;
				break;
		}
		h++;
	}
	citem->next = NULL;
	CreateMenuWindow(&menu, w, h);
	CalculateItemPosition(&menu);
}
/*----------------------------------------------------------------------------*/
void VKDestroyMenuWindow()
{
	vk_menu_item *item = menu.items;
	while( item ) {
		vk_menu_item *mi = item;
		item = item->next;
		if( mi->menu ) {
			VKUnregisterEvent(mi->menu->win);
			XFreeGC(display, mi->menu->gc);
			XDestroyWindow(display, mi->menu->win);
		}
	}
	VKUnregisterEvent(menu.win);
	XFreeGC(display, menu.gc);
	XDestroyWindow(display, menu.win);
}
/*----------------------------------------------------------------------------*/
void VKHideMenu()
{
	if( menu.sidex && menu.sitem->type==0 )
		XUnmapWindow(display, menu.sitem->menu->win);
	XUnmapWindow(display, menu.win);
	vk_menu_on = False;
}
/*----------------------------------------------------------------------------*/
void DrawMenu(vk_menu *m)
{
	vk_menu_item *item = m->items;
#ifdef USE_XFT
	XftDraw *draw;
	draw = XftDrawCreate(display, m->win, visual, colormap);
#endif
	XClearWindow(display, m->win);
	VKSetColor(display, m->gc, clBackground);
	XFillRectangle(display, m->win, m->gc, 0, 0, m->w, m->h);
	VKSetColor(display, m->gc, clMenuBar);
	XDrawRectangle(display, m->win, m->gc, 1, 1, m->w-3, m->h-3);
	XFillRectangle(display, m->win, m->gc, 1, 1, m->w-3, vk_text_height + 4);
#ifdef USE_XFT
	XftDrawStringUtf8(draw, &clBlack, vk_font, m->tx+1, m->ty+1,
							(XftChar8*)m->name, strlen(m->name));
	XftDrawStringUtf8(draw, &clWhite, vk_font, m->tx, m->ty,
							(XftChar8*)m->name, strlen(m->name));
#else
	VKSetColor(display, m->gc, clBlack);
	XmbDrawString(display, m->win, vk_fontset, m->gc, m->tx+1, m->ty+1,
							m->name, strlen(m->name));
	VKSetColor(display, m->gc, clWhite);
	XmbDrawString(display, m->win, vk_fontset, m->gc, m->tx, m->ty,
							m->name, strlen(m->name));
#endif
	while( item ) {
	#ifdef USE_XFT
		XftColor clItem =
	#else
		ulong clItem =
	#endif
			item->type > 0 && item->vptr ?
			(*item->vptr==item->value ? clText : clGray) :
			clBlack;

	#ifdef USE_XFT
		XftDrawStringUtf8(draw, &clItem, vk_font, item->x, item->y,
								(XftChar8*)item->name, item->length);
	#else
		VKSetColor(display, m->gc, clItem);
		XmbDrawString(display, m->win, vk_fontset, m->gc, item->x, item->y,
								item->name, item->length);
	#endif
		if( item->type==0 ) {
			int x = m->w - 15;
			int y = item->y - vk_text_height/2 + 4;
			VKSetColor(display, m->gc, clBlack);
			XDrawLine(display, m->win, m->gc, x, y-1, x+1, y-1);
			XDrawLine(display, m->win, m->gc, x, y, x+3, y);
			XDrawLine(display, m->win, m->gc, x, y+1, x+1, y+1);
		}
		item = (vk_menu_item *)item->next;
	}
#ifdef USE_XFT
	XftDrawDestroy(draw);
#endif
}
/*----------------------------------------------------------------------------*/
void ShowMenu(vk_menu *p, vk_menu *m)
{
	int px, py;
	
	if( p ) {
		px = p->x + p->w;
		py = p->y + (p->sidex - 1) * (vk_text_height + 4) + 2;
	}
	else {
		px = vk_x;
		py = vk_y;
	}

	if( px+m->w > screen_width )
		m->x = (p ? p->x : screen_width) - m->w;
	else
		m->x = px;

	if( py+m->h > screen_height )
		m->y = (p ? 2*vk_text_height+MENU_PADDING+4 : 0) + py - m->h;
	else
		m->y = p ? py-2 : py + vk_icon_height;

	XMoveWindow(display, m->win, m->x, m->y);
	XMapRaised(display, m->win);

	vk_mouse_clicked = False;
	vk_menu_on = True;
}
/*----------------------------------------------------------------------------*/
void VKShowMenu()
{
	ShowMenu(NULL, &menu);
}
/*----------------------------------------------------------------------------*/
void MenuMousePress(vk_menu *m, XEvent *event)
{
	vk_menu_item *item = m->sitem;

	if( m->sidex )
	switch( item->type ) {
		case -1:
			vk_done = True;
			break;
		case -2:
			VKShowFlash();
			break;
		case 1:
			if( vk_method != item->value ) {
				vk_method = item->value;
				if( vk_method != VKM_OFF ) {
					vk_using = vk_method;
					VKChangeUsing();
				}
				VKChangeMethod();
			}
			break;
		case 3:
			if( vk_charset != item->value )
				VKChangeCharset(vk_charset = item->value);
			break;
		case 4:
			if( vk_interface!=item->value ) {
				vk_interface = item->value;
				VKDrawHotkeyWindow();
				VKDestroyMenuWindow();
				VKCreateMenuWindow();
				VKIconLoadInterface();
				VKDestroyFlash();
				VKCreateFlash();
			}
			break;
		case 5:
			VKShowHotkeyWindow();
			break;
	#ifdef VK_CHECK_SPELLING
		case 7:
			vk_spelling = (vk_spelling + 1) & 1;
			VKSetSpelling(vk_spelling);
			break;
	#endif
		case 8:
			if( vk_docking != VKD_ON )
				VKDockMainWindow();
			else
				VKUndockMainWindow();
			break;
	}
	VKDrawIcon();
	VKHideMenu();
}
/*----------------------------------------------------------------------------*/
void MenuMouseMove(vk_menu *m, XEvent *event)
{
	int idx = (event->xbutton.y - MENU_PADDING) / (vk_text_height + 4),
		c = m->sidex, n;

	if( c!=idx ) {
		if( c ) {
			VKSetColor(display, m->gc, clWhite);
			XDrawRectangle(display, m->win, m->gc,
							4, c*(vk_text_height+4)+2,
							m->w-9, vk_text_height+2);
			if( m->sitem->type==0 )
				XUnmapWindow(display, m->sitem->menu->win);
		}
		m->sidex = idx;
		if( idx ) {
			vk_menu_item *mi = m->items;
			VKSetColor(display, m->gc, clBlack);
			XDrawRectangle(display, m->win, m->gc,
							4, idx*(vk_text_height+4)+2,
							m->w-9, vk_text_height+2);
			for( n=1; n<idx && mi; n++ )
				mi = mi->next;
			if( mi ) {
				m->sitem = mi;
				if( mi->type==0 )
					ShowMenu(m, mi->menu);
			}
		}
	}
}
/*----------------------------------------------------------------------------*/
void MenuProcess(XEvent *event, void *data)
{
	vk_menu *m = (vk_menu *)data;
	switch( event->type ) {
		case Expose:
		case GraphicsExpose:
			DrawMenu(m);
			break;
		case ButtonPress:
			MenuMousePress(m, event);
			break;
		case MotionNotify:
			MenuMouseMove(m, event);
			break;
	}
}
/*----------------------------------------------------------------------------*/
void VKUpdateDockingMessage()
{
	vk_menu_item *item = &vk_menu_data[vk_interface][DOCK_MSG_IDX];
	item->name = vk_dock_message[vk_interface][vk_docking == VKD_ON];
	item->length = strlen(item->name);
}
/*----------------------------------------------------------------------------*/
